### PR TITLE
Robustness of the TR minimizers.

### LINKED ISSUE #16771: Improve robustness / efficiency of the trust region minimizers.
Apply a patch of changes to the code that should improve the minimizers' robustness.

### PR DESCRIPTION
**Pull request by mantid-roman**
Applied a patch provided by Tyrone Rees that fixes getting NaNs.

**To test:**

Run the following script. The fit won't be perfect but there won't be any NaNs either.

```
x = [1, 2, 3, 5, 7, 10]
y = [109, 149, 149, 191, 213, 224]
e = [1, 1, 1, 1, 1, 1]

data = CreateWorkspace(x, y, e)
fun = 'name=UserFunction,Formula=b1*(1-exp(-b2*x)),b1=1,b2=1'
Fit(fun, data, Output='fit', Minimizer='DTRS')
```

Change the minimizer to More-Sorensen and run the script again.

Fixes #16771.

_Does not need to be in the release notes._

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by samueljackson92** [comment]
Changes look reasonable. Script no longer produces a runtime error with either minimizer. Looks good to me. :shipit:

File: Framework/CurveFitting/src/FuncMinimizers/TrustRegionMinimizer.cpp [hunk 1]
```diff
@@ -272,10 +272,22 @@ bool TrustRegionMinimizer::iterate(size_t) {
     //
     // if model is good, rho should be close to one
     rho = calculate_rho(w.normF, normFnew, md, options);
-    if (boost::math::isnan(rho) || boost::math::isinf(rho)) {
-      throw std::runtime_error("Rho is NaN or infinite.");
-    }
-    if (rho > options.eta_successful) {
+    if (boost::math::isnan(rho) || boost::math::isinf(rho) ||
```
Comment (quantumsteve) [inline Framework/CurveFitting/src/FuncMinimizers/TrustRegionMinimizer.cpp]: what about [std::isfinite](http://www.cplusplus.com/reference/cmath/isfinite/)?

**Comment by mantid-roman** [comment]
@samueljackson92 Could you take another look at this?

**Comment by samueljackson92** [comment]
@mantid-roman shall do.

**Review by samueljackson92 — APPROVED** [review]
Built and ran the test locally before I left on Friday, but forgot to comment. The new test passed and I already reviewed the code. :shipit:

**Pull request merged**